tempesta_fw/connection.c

```diff
@@ -94,28 +89,15 @@ tfw_connection_release(TfwConnection *conn)
 int
 tfw_connection_send(TfwConnection *conn, TfwMsg *msg)
 {
-	return ss_send(conn->sk, &msg->skb_list, msg->ss_flags);
+	return TFW_CONN_HOOK_CALL(conn, conn_send, msg);
 }
 
 int
 tfw_connection_recv(void *cdata, struct sk_buff *skb, unsigned int off)
 {
 	TfwConnection *conn = cdata;
 
-	if (!conn->msg) {
-		conn->msg = TFW_CONN_HOOK_CALL(conn, conn_msg_alloc);
-		if (!conn->msg) {
-			__kfree_skb(skb);
-			return -ENOMEM;
-		}
-		TFW_DBG("Link new msg %p with connection %p\n",
-			conn->msg, conn);
-	}
-
-	TFW_DBG("Add skb %p to message %p\n", skb, conn->msg);
-	ss_skb_queue_tail(&conn->msg->skb_list, skb);
-
-	return tfw_gfsm_dispatch(&conn->msg->state, conn, skb, off);
+	return tfw_gfsm_dispatch(&conn->state, conn, skb, off);
```

Comment:
Yeah, looks like a very good solution: there is definitely connection state which we didn't have before and which differs from messages state. Also `conn_msg_alloc` seems somewhat unnatural for connectlon layer. So the hook should be removed - now we call `tfw_http_conn_msg_alloc()` from http.c only.